### PR TITLE
Add OAuth 2.0 Interceptor

### PR DESCRIPTION
**Pull request by dijalmasilva**
Adding new feature to provide OAuth 2.0 in Heimdall. Reference Issue #17

### CODE REVIEW

File: heimdall-core/src/main/java/br/com/conductor/heimdall/core/dto/ProviderDTO.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import lombok.Data;
+
+/**
+ * @author <a href="https://dijalmasilva.github.io" target="_blank">Dijalma Silva</a>
+ *
+ */
+@Data
+public class ProviderDTO implements Serializable{
+
+	private static final long serialVersionUID = 6270561769116836980L;
```
Comment (MAAARKIN) [inline heimdall-core/src/main/java/br/com/conductor/heimdall/core/dto/ProviderDTO.java]: indentation problem?

File: heimdall-core/src/main/java/br/com/conductor/heimdall/core/entity/Provider.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+ */
+@Data
+@Table(name = "PROVIDERS")
+@Entity
+@DynamicUpdate
+@DynamicInsert
+@EqualsAndHashCode(of = { "id" })
+public class Provider implements Serializable {
+
+	private static final long serialVersionUID = 6667573637255450718L;
```
Comment (MAAARKIN) [inline heimdall-core/src/main/java/br/com/conductor/heimdall/core/entity/Provider.java]: again?

File: heimdall-core/src/main/java/br/com/conductor/heimdall/core/entity/ProviderParam.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+ */
+@Data
+@Table(name = "PROVIDER_PARAMS")
+@Entity
+@DynamicUpdate
+@DynamicInsert
+@EqualsAndHashCode(of = { "id" })
+public class ProviderParam implements Serializable{
+
+	private static final long serialVersionUID = 2575009691412505853L;
```
Comment (MAAARKIN) [inline heimdall-core/src/main/java/br/com/conductor/heimdall/core/entity/ProviderParam.java]: indentation

File: heimdall-core/src/main/java/br/com/conductor/heimdall/core/util/JwtUtils.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		TokenOAuth tokenOAuth = new TokenOAuth();
+		tokenOAuth.setAccessToken(accessToken);
+		tokenOAuth.setRefreshToken(refreshToken);
+		tokenOAuth.setExpiration(LocalDateTime.now().until(dateExpiredAccessToken, ChronoUnit.SECONDS));
+		return tokenOAuth;
+	}
+	
+	private Claims getClaimsFromTheToken(String token, String secretKey) throws Exception {
+		 
+		Claims claims;
```
Comment (MAAARKIN) [inline heimdall-core/src/main/java/br/com/conductor/heimdall/core/util/JwtUtils.java]: indentation...

File: heimdall-gateway/src/main/java/br/com/conductor/heimdall/gateway/service/InterceptorFileService.java [hunk 5]
```diff
@@ -223,6 +225,9 @@ private File templateInterceptor(TypeInterceptor type, TypeExecutionPoint execut
                     case MIDDLEWARE:
                          file = ResourceUtils.getFile(filePath + File.separator + "middleware.mustache");
                          break;
+                    case OAUTH:
+        				file = ResourceUtils.getFile(filePath + File.separator + "oauth.mustache");
```
Comment (MAAARKIN) [inline heimdall-gateway/src/main/java/br/com/conductor/heimdall/gateway/service/InterceptorFileService.java]: indentation...

File: heimdall-gateway/src/main/java/br/com/conductor/heimdall/gateway/service/InterceptorFileService.java [hunk 6]
```diff
@@ -388,6 +393,16 @@ private Integer definePrefixOrder(InterceptorLifeCycle lifeCycle) {
                     parameters.put("interval", rateLimitDTO.getInterval().name());
                }
 
+               if (objectCustom instanceof OAuthDTO) {
+
+					OAuthDTO oAuthDTO = (OAuthDTO) objectCustom;
```
Comment (MAAARKIN) [inline heimdall-gateway/src/main/java/br/com/conductor/heimdall/gateway/service/InterceptorFileService.java]: again...

File: heimdall-core/src/main/java/br/com/conductor/heimdall/core/entity/ProviderParam.java (hunk 3 above)
Comment (dijalmasilva) [reply heimdall-core/src/main/java/br/com/conductor/heimdall/core/entity/ProviderParam.java]: fixed.

**Comment by MAAARKIN** [comment]
@dijalmasilva plz, remember to put javadoc.

**Pull request merged**